File: src/Leitor/LeitorDocente.java
package Leitor;

import java.util.Scanner;
import java.util.Set;
import java.io.Serializable;
import java.util.HashMap;

import Modelo.Docente;

public class LeitorDocente extends ILeitor implements Serializable{
   
    private static final long serialVersionUID = 1L;
    private HashMap<String, Docente> mapa = new HashMap<>();
    private static LeitorDocente leitor;

    private LeitorDocente(){}

    public static LeitorDocente obterInstancia(){
        if (leitor == null){
            leitor = new LeitorDocente();
            return leitor;
        }
        else return leitor;
    }

    public HashMap<String,Docente> obterHash(){
        return mapa;
    }

    public static void setInstancia(LeitorDocente novo){
        leitor = novo;
    }

    public Docente ler(Scanner scanner){
        String login = scanner.next();
        if(leitor.busca(login) != null)
			throw new IllegalArgumentException("Cadastro repetido: "+login+".");
        String nome = scanner.next();
        Docente docente;
        String site = scanner.next();
        docente = new Docente(login,nome,site);
        
        return docente;
    }

    public void anexaHash(Docente docente){
        mapa.put(docente.obterLogin(), docente);
    }

    public Docente busca(String login){
        if(mapa.containsKey(login)){
            return mapa.get(login);
        }
        return null;
    }

    public Set<String> obterChaves(){
        return mapa.keySet();
    }

}

File: cpp_version/Leitor/LeitorDocente.cpp
#include "LeitorDocente.h"

LeitorDocente* LeitorDocente::leitor = nullptr;

LeitorDocente::LeitorDocente(){}

LeitorDocente* LeitorDocente::obterInstancia(){
    if (leitor == NULL){
        leitor = new LeitorDocente();
        return leitor;
    }
    else return leitor;
}

vector<Docente*> LeitorDocente::obterDocentes(){
    vector<Docente*> lista;
    for(map<string,Docente*>::iterator it = mapDocente.begin(); it != mapDocente.end(); ++it){
        lista.push_back(it->second);
    }
    return lista;
}

Docente* LeitorDocente::ler(vector<string> vec){
    string login = vec[0];
    
    if(leitor->busca(login) != NULL)
        throw ExcecaoCad("Cadastro repetido: " + login + ".");

    string nome = vec[1];
    string site = vec[2];

    return new Docente(login, nome, site);
}

void LeitorDocente::anexaHash(Docente* docente){
    mapDocente[docente->obterLogin()] = docente;
}

Docente* LeitorDocente::busca(string login){
    bool contains = false;
    for(map<string,Docente*>::iterator it = mapDocente.begin(); it != mapDocente.end(); ++it){
        if(it->first == login){
            contains = true;
        }
    }
    if(contains){
        return mapDocente[login];
    }
    return nullptr;
}

vector<Periodo*> LeitorDocente::busca(Docente* docente){
    ControladorDisciplina cDisciplina;

    vector<Disciplina*> disciplinas = cDisciplina.busca(docente);
    vector<Periodo*> periodos;

    bool possui = false;
    for(Disciplina* disciplina : disciplinas){
        possui = false;
        for(Periodo* periodo: periodos){
            if(disciplina->obterPeriodo() == periodo){
                possui=true;
            }
        }
        if(!possui){
            periodos.push_back(disciplina->obterPeriodo());
        }
    }
    return periodos;
}